File: articles/life-death-problem-of-biological-definition/r/dormancy_model.R
# Dormancy loss and reactivation model in R.

scenario_path <- file.path("data", "dormancy_scenarios.csv")

if (!file.exists(scenario_path)) {
  scenario_path <- file.path("..", "data", "dormancy_scenarios.csv")
}

simulate_dormancy <- function(dormant_initial, active_initial, mortality_rate, reactivation_rate, time_end, dt) {
  time <- seq(0, time_end, by = dt)

  dormant <- numeric(length(time))
  active <- numeric(length(time))
  dead_or_lost <- numeric(length(time))

  dormant[1] <- dormant_initial
  active[1] <- active_initial
  dead_or_lost[1] <- 0

  for (i in 2:length(time)) {
    d_dormant <- -(mortality_rate + reactivation_rate) * dormant[i - 1]
    d_active <- reactivation_rate * dormant[i - 1]
    d_lost <- mortality_rate * dormant[i - 1]

    dormant[i] <- max(dormant[i - 1] + d_dormant * dt, 0)
    active[i] <- active[i - 1] + d_active * dt
    dead_or_lost[i] <- dead_or_lost[i - 1] + d_lost * dt
  }

  data.frame(
    time = time,
    dormant = dormant,
    active = active,
    dead_or_lost = dead_or_lost
  )
}

scenarios <- read.csv(scenario_path)

rows <- lapply(seq_len(nrow(scenarios)), function(i) {
  s <- scenarios[i, ]

  sim <- simulate_dormancy(
    s$dormant_initial,
    s$active_initial,
    s$mortality_rate,
    s$reactivation_rate,
    s$time_end,
    s$dt
  )

  initial_total <- s$dormant_initial + s$active_initial

  data.frame(
    scenario = s$scenario,
    final_dormant = tail(sim$dormant, 1),
    final_active = tail(sim$active, 1),
    final_dead_or_lost = tail(sim$dead_or_lost, 1),
    retained_viable_fraction = (tail(sim$dormant, 1) + tail(sim$active, 1)) / initial_total,
    activated_fraction = tail(sim$active, 1) / initial_total
  )
})

summary_df <- do.call(rbind, rows)

print(round(summary_df, 5))

File: articles/life-death-problem-of-biological-definition/python/dormancy_model.py
"""
Dormancy loss and reactivation workflow.

Run:
    python python/dormancy_model.py
"""

from __future__ import annotations

from pathlib import Path

import numpy as np
import pandas as pd

from definition_core import simulate_dormancy


ARTICLE_DIR = Path(__file__).resolve().parents[1]
DORMANCY_PATH = ARTICLE_DIR / "data" / "dormancy_scenarios.csv"


def main() -> None:
    """Simulate dormancy scenarios."""

    scenarios = pd.read_csv(DORMANCY_PATH)

    rows = []

    for _, scenario in scenarios.iterrows():
        time = np.arange(0, scenario["time_end"] + scenario["dt"], scenario["dt"])

        sim = simulate_dormancy(
            time=time,
            dormant_initial=scenario["dormant_initial"],
            active_initial=scenario["active_initial"],
            mortality_rate=scenario["mortality_rate"],
            reactivation_rate=scenario["reactivation_rate"],
        )

        initial_total = scenario["dormant_initial"] + scenario["active_initial"]

        rows.append(
            {
                "scenario": scenario["scenario"],
                "final_dormant": sim["dormant"][-1],
                "final_active": sim["active"][-1],
                "final_dead_or_lost": sim["dead_or_lost"][-1],
                "retained_viable_fraction": (sim["dormant"][-1] + sim["active"][-1]) / initial_total,
                "activated_fraction": sim["active"][-1] / initial_total,
            }
        )

    print(pd.DataFrame(rows).round(5).to_string(index=False))


if __name__ == "__main__":
    main()
